src/Disks/DiskWebServer.cpp

```diff
@@ -125,204 +112,149 @@ class ReadBufferFromWebServer final : public ReadIndirectBufferFromRemoteFS<Read
     }
 
 private:
     String uri;
     ContextPtr context;
-    size_t max_read_tries;
     size_t buf_size;
 };
 
 
-class WriteBufferFromNothing : public WriteBufferFromFile
-{
-public:
-    WriteBufferFromNothing() : WriteBufferFromFile("/dev/null") {}
-
-    void sync() override {}
-};
-
-
 DiskWebServer::DiskWebServer(
             const String & disk_name_,
-            const String & uri_,
-            const String & metadata_path_,
+            const String & url_,
             ContextPtr context_,
-            SettingsPtr settings_)
+            size_t min_bytes_for_seek_)
         : WithContext(context_->getGlobalContext())
         , log(&Poco::Logger::get("DiskWeb"))
-        , uri(uri_)
+        , url(url_)
         , name(disk_name_)
-        , metadata_path(metadata_path_)
-        , settings(std::move(settings_))
+        , min_bytes_for_seek(min_bytes_for_seek_)
 {
 }
 
 
-String DiskWebServer::getFileName(const String & path) const
+bool DiskWebServer::exists(const String & path) const
 {
-    String result;
-
-    if (RE2::FullMatch(path, MATCH_DIRECTORY_FILE_PATTERN)
-        && RE2::Extract(path, MATCH_DIRECTORY_FILE_PATTERN, fmt::format(R"({}-\1-\2-\3)", settings->files_prefix), &result))
-        return result;
-
-    if (RE2::FullMatch(path, MATCH_ROOT_FILE_PATTERN)
-        && RE2::Extract(path, MATCH_ROOT_FILE_PATTERN, fmt::format(R"({}-\1-\2)", settings->files_prefix), &result))
-        return result;
-
-    throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected file: {}", path);
-}
-
+    LOG_TRACE(&Poco::Logger::get("DiskWeb"), "Checking existence of path: {}", path);
 
-bool DiskWebServer::findFileInMetadata(const String & path, File & file_info) const
-{
-    String table_uuid, directory_name, file_name;
+    if (files.find(path) != files.end())
+        return true;
 
-    if (RE2::FullMatch(path, MATCH_DIRECTORY_FILE_PATTERN, &table_uuid, &directory_name, &file_name)
-       || RE2::FullMatch(path, MATCH_ROOT_FILE_PATTERN, &table_uuid, &file_name)
-       || RE2::FullMatch(path, MATCH_DIRECTORY_PATTERN, &table_uuid, &directory_name))
+    if (path.ends_with(MergeTreeData::FORMAT_VERSION_FILE_NAME) && files.find(fs::path(path).parent_path() / "") == files.end())
     {
-        if (directory_name.empty())
-            directory_name = file_name;
-
-        if (!metadata.tables_data.count(table_uuid))
-            return false;
-
-        if (!metadata.tables_data[table_uuid].count(directory_name))
-            return false;
-
-        if (file_name.empty())
-            return true;
+        try
+        {
+            initialize(fs::path(url) / fs::path(path).parent_path());
+            return files.find(path) != files.end();
+        }
+        catch (...)
+        {
+            const auto message = getCurrentExceptionMessage(false);
+            bool can_throw = CurrentThread::isInitialized() && CurrentThread::get().getQueryContext();
+            if (can_throw)
```

Comment:
@kssenii why it cannot throw if it is loaded from the server context? (read server start) This makes the table simply empty, which is not good:

```
2022.10.28 15:18:30.739698 [ 10663 ] {} <Error> ReadWriteBufferFromHTTP: HTTP request to `http://127.0.0.1:8080/store/a3e/a3e65e1f-5fd4-47ed-9dbd-307f2586b52d/detached/.index` failed at try 1/1 with bytes read: 0/unknown. Error: Connection refused. (Current backoff wait is 100/1600 ms)
2022.10.28 15:18:30.841210 [ 10663 ] {} <Trace> DiskWeb: Cannot load disk metadata. Error: Poco::Exception. Code: 1000, e.code() = 111, Connection refused (version 22.11.1.1                                                       
```

```sql
SELECT *
FROM data_from_web

Query id: ee544a5e-3c67-4fb4-8f14-f8e4a082b237

Ok.

0 rows in set. Elapsed: 0.019 sec. 
```

Reply:
@azat this is clearly a wrong behavior, and should be fixed.

Reply:
https://github.com/ClickHouse/ClickHouse/pull/42800

Reply:
This was done in order to allow disk Web to start even if there is no connection on server startup.